### PR TITLE
Adding Proxy Support to hltGetConfiguration and fixing converter/db specification bugs with hltIntegrationTests

### PR DESCRIPTION
**Pull request by Sam-Harper**
#### PR description:

The PR updates hltGetConfiguration and friends to be able to use a socks proxy thus simplifying hlt work outside the cern network. 

It also fixes a bug in hltIntegrationTests reported by Andres Leonardo Cabrera Mora which has to do with how the database/converter is selected for the "setup menu" download. 

--dbproxy will now activate this and --dbproxyhost/--dbproxyport will allow hosts/ports other than localhost:8080

Note this PR is a work in progress, the main v3 converter is not updated (only v3-test)  and Andrea just made a good suggestion about alternatives to sha512 hashes which I would like explore. However I thought it would be nice to have the rest of the PR available to also receive comments rather than waiting to explore that suggestion. 

#### PR validation:
```
 hltIntegrationTests v3-test/run3:/dev/CMSSW_12_0_0/Fake -s /dev/CMSSW_12_0_0/GRun -i /store/data/Run2018D/EphemeralHLTPhysics1/RAW/v1/000/323/790/00000/15FD6507-9B9A-DE49-973B-B63C4DEDC226.root -x  "--globaltag auto:run3_hlt" --dbproxy 
hltGetConfiguration v3-test/run3:/dev/CMSSW_12_0_0/Fake --dbproxy
hltConfigFromDB --configName /dev/CMSSW_12_0_0/GRun/V1  --dbproxy --v3-test --run3
 ```

commands all work as expected at RAL (ie very much off the cern network)

### CODE REVIEW

**Comment by missirol** [comment]
Related updates in `ConfDB` are in https://github.com/cms-sw/hlt-confdb/pull/34.

In my understanding, the PR tests in CMSSW won't be sensitive to any of these updates (except possible python-syntax errors, which I checked are absent); if so, the tests can be triggered later on, once the PR has converged.

Also verified that the following now works outside the CERN network. Thanks, Sam!
```bash
ssh -f -N -D 8081 missirol@lxplus.cern.ch
hltGetConfiguration v3-test/run3:/dev/CMSSW_12_0_0/Fake --dbproxy --dbproxyport 8081
```

File: HLTrigger/Configuration/scripts/hltGetConfiguration [hunk 1]
```diff
@@ -76,6 +76,24 @@ parser.add_argument('--l1-emulator',
                     const   = 'Full',
                     help    = 'Run the Full stage-2 L1T emulator.' )
 
+parser.add_argument('--dbproxy',
+                    dest    = 'proxy',
+                    action  = 'store_const',
+		    const   = 'True',                    
+                    help    = 'uses a socks proxy to connect outside CERN network' )
```
Comment (missirol) [inline HLTrigger/Configuration/scripts/hltGetConfiguration:83]:
```suggestion
                    const   = 'True',                    
                    help    = 'Use a socks proxy to connect outside CERN network (default: False)' )
```

File: HLTrigger/Configuration/scripts/hltGetConfiguration [hunk 2]
```diff
@@ -76,6 +76,24 @@ parser.add_argument('--l1-emulator',
...
+                    action  = 'store',
+                    metavar = 'PROXYPORT',
+                    default = defaults.proxy_port,
+                    help    = 'port for socks proxy' )
+parser.add_argument('--dbproxyhost',
+                    dest    = 'proxy_host',
+                    action  = 'store',
+                    metavar = 'PROXYPORT',
+                    default = defaults.proxy_host,
+                    help    = 'host for socks proxy' )
```
Comment (missirol) [inline HLTrigger/Configuration/scripts/hltGetConfiguration:95]:
```suggestion
                    help    = 'Host of the socks proxy (default: "localhost")' )
```

File: HLTrigger/Configuration/scripts/hltGetConfiguration [hunk 3]
```diff
@@ -76,6 +76,24 @@ parser.add_argument('--l1-emulator',
...
+parser.add_argument('--dbproxyport',
+                    dest    = 'proxy_port',
+                    action  = 'store',
+                    metavar = 'PROXYPORT',
+                    default = defaults.proxy_port,
+                    help    = 'port for socks proxy' )
+parser.add_argument('--dbproxyhost',
+                    dest    = 'proxy_host',
+                    action  = 'store',
+                    metavar = 'PROXYPORT',
```
Comment (missirol) [inline HLTrigger/Configuration/scripts/hltGetConfiguration:93]:
```suggestion
                    metavar = 'PROXYHOST',
```

File: HLTrigger/Configuration/scripts/hltGetConfiguration [hunk 4]
```diff
@@ -76,6 +76,24 @@ parser.add_argument('--l1-emulator',
...
+                    dest    = 'proxy',
+                    action  = 'store_const',
+		    const   = 'True',                    
+                    help    = 'uses a socks proxy to connect outside CERN network' )
+parser.add_argument('--dbproxyport',
+                    dest    = 'proxy_port',
+                    action  = 'store',
+                    metavar = 'PROXYPORT',
+                    default = defaults.proxy_port,
+                    help    = 'port for socks proxy' )
```
Comment (missirol) [inline HLTrigger/Configuration/scripts/hltGetConfiguration:89]:
```suggestion
                    help    = 'Port of the socks proxy (default: 8080)' )
```

File: HLTrigger/Configuration/scripts/hltListPaths [hunk 5]
```diff
@@ -55,6 +60,24 @@ parser.add_argument('menu',
                     type    = options.ConnectionHLTMenu,
                     metavar = 'MENU', 
                     help    = 'HLT menu to dump from the database. Supported formats are:\n  - /path/to/configuration[/Vn]\n  - [[{v1|v2|v3}/]{run3|run2|online|adg}:]/path/to/configuration[/Vn]\n  - run:runnumber\nThe possible converters are "v1", "v2, and "v3" (default).\nThe possible databases are "run3" (default, used for offline development), "run2" (used for accessing run2 offline development menus), "online" (used to extract online menus within Point 5) and "adg" (used to extract the online menus outside Point 5).\nIf no menu version is specified, the latest one is automatically used.\nIf "run:" is used instead, the HLT menu used for the given run number is looked up and used.\nNote other converters and databases exist as options but they are only for expert/special use.' )
+parser.add_argument('--dbproxy',
+                    dest    = 'proxy',
+                    action  = 'store_const',
+		    const   = 'True',                    
+                    help    = 'uses a socks proxy to connect outside CERN network' )
```
Comment (missirol) [inline HLTrigger/Configuration/scripts/hltListPaths:67]:
```suggestion
                    const   = 'True',                    
                    help    = 'Use a socks proxy to connect outside CERN network (default: False)' )
```

File: HLTrigger/Configuration/scripts/hltListPaths [hunk 6]
```diff
@@ -55,6 +60,24 @@ parser.add_argument('menu',
...
+parser.add_argument('--dbproxyport',
+                    dest    = 'proxy_port',
+                    action  = 'store',
+                    metavar = 'PROXYPORT',
+                    default = defaults.proxy_port,
+                    help    = 'port for socks proxy' )
+parser.add_argument('--dbproxyhost',
+                    dest    = 'proxy_host',
+                    action  = 'store',
+                    metavar = 'PROXYPORT',
```
Comment (missirol) [inline HLTrigger/Configuration/scripts/hltListPaths:77]:
```suggestion
                    metavar = 'PROXYHOST',
```

File: HLTrigger/Configuration/scripts/hltListPaths [hunk 7]
```diff
@@ -55,6 +60,24 @@ parser.add_argument('menu',
...
+                    default = defaults.proxy_host,
+                    help    = 'host for socks proxy' )
```
Comment (missirol) [inline HLTrigger/Configuration/scripts/hltListPaths:79]:
```suggestion
                    help    = 'Host of the socks proxy (default: "localhost")' )
```

File: HLTrigger/Configuration/scripts/hltListPaths [hunk 8]
```diff
@@ -55,6 +60,24 @@ parser.add_argument('menu',
...
+                    dest    = 'proxy',
+                    action  = 'store_const',
+		    const   = 'True',                    
+                    help    = 'uses a socks proxy to connect outside CERN network' )
+parser.add_argument('--dbproxyport',
+                    dest    = 'proxy_port',
+                    action  = 'store',
+                    metavar = 'PROXYPORT',
+                    default = defaults.proxy_port,
+                    help    = 'port for socks proxy' )
```
Comment (missirol) [inline HLTrigger/Configuration/scripts/hltListPaths:73]:
```suggestion
                    help    = 'Port of the socks proxy (default: 8080)' )
```

**Review by missirol — COMMENTED** [review]
Minor comments, mainly on output of `--help`.

**Comment by Sam-Harper** [comment]
thanks Marino for the review and independent test, I'll implement those fixes. I've pushed a slight rework along Andreas suggestion to https://github.com/cms-sw/hlt-confdb/pull/34 so it uses the git hash so I'll update this PR to work with that as well.

**Comment by Sam-Harper** [comment]
okay I adjusted it to read the version and  also the v3 is now at V03-00-08 so this is ready to go there

**Comment by missirol** [comment]
please test

(will do some more manual checking on Monday)

File: HLTrigger/Configuration/python/Tools/confdbOfflineConverter.py [hunk 9]
```diff
@@ -94,27 +119,38 @@ def __init__(self, version = 'v3', database = 'run3', url = None, verbose = Fals
             self.workDir = self.baseDir
         else:
             # try to use $CMSSW_BASE/tmp
-            self.workDir = OfflineConverter.CheckTempDirectory(os.environ['CMSSW_BASE'] + '/tmp/confdb')
+            self.workDir = OfflineConverter.CheckTempDirectory(os.path.join(os.environ['CMSSW_BASE'],'tmp','confdb',self.version))
             if not self.workDir:
                 # try to use $TMP
-                self.workDir = OfflineConverter.CheckTempDirectory(os.environ['TMP'] + '/confdb')
+                self.workDir = OfflineConverter.CheckTempDirectory(os.path.join(os.environ['TMP'] + 'confdb',self.version))
```
Comment (missirol) [inline HLTrigger/Configuration/python/Tools/confdbOfflineConverter.py:125]:
```suggestion
                self.workDir = OfflineConverter.CheckTempDirectory(os.path.join(os.environ['TMP'],'confdb',self.version))
```
Or is it intentional to remove the `/` before `confdb`?

File: HLTrigger/Configuration/scripts/hltIntegrationTests [hunk 10]
```diff
@@ -44,6 +44,9 @@ Usage:
        --threads  THREADS     Run with THREADS threads when running the whole HLT (default 4)
   -x | --extra    OPTIONS     Pass OPTIONS directly to hltGetConfiguration
        --mc                   Run over MC instead of data (the default)
+  --dbproxy                   Use a socks proxy to connect to the HLT database
+  --dbproxyhost               Host of the socks proxy (default localhost)
+  --dbproxyport               Port of the socks proxy (default 8080)
```
Comment (missirol) [inline HLTrigger/Configuration/scripts/hltIntegrationTests:49]:
```suggestion
  --dbproxyhost    PROXYHOST  Host of the socks proxy (default: "localhost")
  --dbproxyport    PROXYPORT  Port of the socks proxy (default: 8080)
```

**Comment by missirol** [comment]
Testing the different scripts outside the CERN network worked as expected for me.
I left a couple of final comments; apologies for not sending those earlier.

File: HLTrigger/Configuration/python/Tools/confdbOfflineConverter.py (hunk 9 above)
Comment (Sam-Harper) [reply HLTrigger/Configuration/python/Tools/confdbOfflineConverter.py:125]: no an error, thanks!

**Comment by Sam-Harper** [comment]
thanks Marino, I just added them (I do love this github feature)

**Comment by missirol** [comment]
please test

re-rerunning the tests only for completeness (they are not really sensitive to these updates)

as written above, no further comments on this PR from my side

**Comment by missirol** [comment]
please test with #36021

looks like this other unrelated PR is necessary to fix some of the PR tests

**Comment by missirol** [comment]
+hlt

 - related to `ConfDB` updates in cms-sw/hlt-confdb#34

 - despite the label, this PR did not require any externals
   (another PR was needed to work around an unrelated issue in PR tests)

**Comment by qliphy** [comment]
+1

**Comment by qliphy** [comment]
@Sam-Harper   The HLT Validation shows some error in IB CMSSW_12_2_X_2021-11-09-1100 
https://cmssdt.cern.ch/SDT/jenkins-artifacts/HLT-Validation/CMSSW_12_2_X_2021-11-09-1100/slc7_amd64_gcc900/HLT_Integration_GRun_DATA.log

stty: standard input: Inappropriate ioctl for device
Traceback (most recent call last):
  File "/pool/condor/dir_13577/jenkins/workspace/ib-run-HLT/CMSSW_12_2_X_2021-11-09-1100/bin/slc7_amd64_gcc900/hltListPaths", line 108, in <module>
    paths  = getPathList(config)
  File "/pool/condor/dir_13577/jenkins/workspace/ib-run-HLT/CMSSW_12_2_X_2021-11-09-1100/bin/slc7_amd64_gcc900/hltListPaths", line 23, in getPathList
    if config.proxy:
AttributeError: 'Namespace' object has no attribute 'proxy'


It seems related to this PR?  Would you please have a look?

**Comment by qliphy** [comment]
Ok, I noticed there is already a fix merged:  https://github.com/cms-sw/cmssw/pull/36059
Let's see the outcome in next IB.

**Comment by missirol** [comment]
Hi @qliphy, unfortunately the fix in #36059 is unrelated to this.
I have a fix for this error, I can make a PR shortly. Also, we'll need to backport this latest fix to `12_1` and `12_0`.

**Comment by qliphy** [comment]
> Hi @qliphy, unfortunately the fix in #36059 is unrelated to this. I have a fix for this error, I can make a PR shortly. Also, we'll need to backport this latest fix to `12_1` and `12_0`.

Thanks @missirol

**Pull request merged**